keras_cv/layers/preprocessing/mosaic.py

```diff
@@ -0,0 +1,293 @@
+# Copyright 2022 The KerasCV Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+
+from keras_cv import bounding_box
+from keras_cv.layers.preprocessing.base_image_augmentation_layer import (
+    BaseImageAugmentationLayer,
+)
+from keras_cv.utils import preprocessing
+
+
+@tf.keras.utils.register_keras_serializable(package="keras_cv")
+class Mosaic(BaseImageAugmentationLayer):
+    """Mosaic implements the mosaic data augmentation technique.
+
+    Mosaic data augmentation first takes 4 images from the batch and makes a grid.
+    After that based on the offset, a crop is taken to form the mosaic image. Labels
+    are in the same ratio as the the area of their images in the output image. Bounding
+    boxes are translated according to the position of the 4 images.
+
+    Args:
+        offset: A tuple of two floats, a single float or `keras_cv.FactorSampler`.
+            `offset` is used to determine the offset of the mosaic center from the
+            top-left corner of the mosaic. If a tuple is used, the x and y coordinates
+            of the mosaic center are sampled between the two values for every image
+            augmented. If a single float is used, a value between `0.0` and the passed
+            float is sampled.  In order to ensure the value is always the same, please
+            pass a tuple with two identical floats: `(0.5, 0.5)`. Defaults to
+            (0.25, 0.75).
+        bounding_box_format: a case-insensitive string (for example, "xyxy") to be
+            passed if bounding boxes are being augmented by this layer.
+            Each bounding box is defined by at least these 4 values. The inputs
+            may contain additional information such as classes and confidence after
+            these 4 values but these values will be ignored and returned as is. For
+            detailed information on the supported formats, see the
+            [KerasCV bounding box documentation](https://keras.io/api/keras_cv/bounding_box/formats/).
+            Defualts to None.
+        seed: Integer. Used to create a random seed.
+
+    References:
+        - [Yolov4 paper](https://arxiv.org/pdf/2004.10934).
+        - [Yolov5 implementation](https://github.com/ultralytics/yolov5).
+        - [YoloX implementation](https://github.com/Megvii-BaseDetection/YOLOX)
+
+    Sample usage:
+    ```python
+    (images, labels), _ = tf.keras.datasets.cifar10.load_data()
```

Comment:
we can also use keras_cv.datasets to show how to do this for pascalVOC

Reply:
That would make sense. Do you think updating the examples/layers/preprocessing/bounding_box/demo_utils.py would be a better idea here?